Documentation and preconditioner fixes for MFEM backend

nmnobre opened the pull request:
## Reason
Incomplete or incorrectly rendering documentation.
A bug introduced in #30591 prevents the correct selection of the solver's preconditioner.

## Design
Added or fixed documentation. Had to improve `extractInputBlocks()` in `listing.py` to strip trailing slashes only.
Fixed preconditioning selection bug by essentially rolling back to the original design whereby we get the respective user object from the solver constructor, not an `MFEMProblem` method.

## Impact
Beautiful documentation.
Most notably, this should mean we now use the preconditioner we intended to: some tests marked LOR were in fact running without LOR.

nmnobre committed:
Fix input file syntax to show Kernels not AuxKernels (#30660)

nmnobre committed:
Fix MFEM(Vector)L2Error class descriptions (#30660)

nmnobre committed:
Fix preconditioner selection (#30660)

nmnobre committed:
Fix missing BC class descriptions (#30660)

nmnobre committed:
Revert kernels/outputs test arguments (#30660)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2916327/), step Clang format on 75912e8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30660/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30660/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6c73562a4736bcd98b127409ae2cd09dbe138a53`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2916564/), step Docs: sync website on 14ab901 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30660/site)

This comment will be updated on new commits.

nmnobre committed:
Tweak listing block extraction to allow leading slash (#30660)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2917589/), step Versioner verify on 26261c0 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner influential files

Found 57 influential files, 1 changed, 0 added, 0 removed

| package   | status   | file                  |
|-----------|----------|-----------------------|
| wasp      | CHANGE   | python/pyhit/pyhit.py |

### Versioner versions

Found 10 packages, 2 changed, 2 failed

| package   | status    | hash    | to hash   | version            | to version         |
|-----------|-----------|---------|-----------|--------------------|--------------------|
| wasp      | NEED BUMP | 43d14da | a74b6be   | 2025.05.13 build 0 | 2025.05.13 build 0 |
| moose-dev | NEED BUMP | 92d3c2e | ad9b40d   | 2025.05.23         | 2025.05.23         |

Verification failed.

nmnobre commented:
> ### Versioner influential files
> Found 57 influential files, 1 changed, 0 added, 0 removed
> 
> package	status	file
> wasp	CHANGE	python/pyhit/pyhit.py

For clarity, I'm no longer modifying this file. It was just a typo fix in a docstring, and I can't be bothered to bump all the conda stuff just for that...

cticenhour commented on python/MooseDocs/extensions/listing.py:
```diff
@@ -403,7 +403,7 @@ def extractInputBlocks(filename, blocks):
         out = []
         for block in blocks.split():
-            node = moosetree.find(hit, lambda n: n.fullpath.endswith(block.strip('/')))
+            node = moosetree.find(hit, lambda n: n.fullpath.endswith(block.rstrip('/')))
```
I am fine with this extension adjustment. 👍🏻 Will leave other aspects to @lindsayad since he is more familiar with the MFEM integration.

lindsayad commented on framework/include/mfem/solvers/MFEMSolverBase.h:
```diff
@@ -23,7 +23,7 @@ class MFEMSolverBase : public MFEMGeneralUserObject
   virtual void updateSolver(mfem::ParBilinearForm & a, mfem::Array<int> & tdofs) = 0;
 
-  bool isLOR() const { return _lor; }
+  bool isLOR() const { return _lor || (_preconditioner ? _preconditioner->isLOR() : false); }
```
```suggestion
  bool isLOR() const { return _lor || (_preconditioner && _preconditioner->isLOR()); 
```

lindsayad replied:
Some doxygen for why we need this new `||` would be great

nmnobre replied:
Oh that's embarrassing, good catch!

lindsayad commented on framework/src/mfem/solvers/MFEMCGSolver.C:
```diff
@@ -40,8 +36,12 @@ MFEMCGSolver::constructSolver(const InputParameters &)
+  if (isParamSetByUser("preconditioner"))
+  {
+    auto & pre_ref = const_cast<MFEMSolverBase &>(getUserObject<MFEMSolverBase>("preconditioner"));
+    _preconditioner = std::dynamic_pointer_cast<MFEMSolverBase>(pre_ref.getSharedPtr());
```
Do we need `_preconditioner` to be a shared pointer? Would just a regular pointer suffice? Then we can avoid bouncing up and down the hierarchy. MFEMProblem will definitely destruct after this object

nmnobre replied:
I just kept it as it was originally, but I agree it seems unnecessary for `_preconditioner`, so that's now a raw pointer.
I've left `_solver` as a shared pointer though, as changing that in the same way isn't (unless I'm missing something) quite as straightforward, we'd need to manage that memory ourselves.

lindsayad commented on framework/src/mfem/solvers/MFEMHyprePCG.C:
```diff
@@ -41,8 +37,10 @@ MFEMHyprePCG::constructSolver(const InputParameters &)
   solver->SetPrintLevel(getParam<int>("print_level"));
 
-  if (_preconditioner)
+  if (isParamSetByUser("preconditioner"))
   {
+    auto & pre_ref = const_cast<MFEMSolverBase &>(getUserObject<MFEMSolverBase>("preconditioner"));
+    _preconditioner = std::dynamic_pointer_cast<MFEMSolverBase>(pre_ref.getSharedPtr());
```
This pattern repeats over and over for these Solvers. You get the twist of the following four lines for Hypre solvers but that doesn't actually change the implementation. Those are just for error checking (which is a good thing). It seems like maybe we could add something like this:
```c++
void
MFEMSolverBase::finishSolverConstruction(mfem::Solver & solver, const bool check_for_hypre)
{
  if (isParamSetByUser("preconditioner"))
  {
    _preconditioner = &const_cast<MFEMSolverBase &>(getUserObject<MFEMSolverBase>("preconditioner"));
    if (check_for_hypre)
    {
      auto * const hypre_pre = dynamic_cast<mfem::HypreSolver *>(_preconditioner->getSolver())
      if (!hypre_pre)
        mooseError("Preconditioner must be a hypre solver);
    }

    solver.SetPreconditioner(*_preconditioner->getSolver());
  }

  _solver = &solver;
}
```
This method assumes that both `_preconditioner` and `_solver` become regular pointers instead of shared pointers

nmnobre replied:
"Seems" being the keyword. :) Unfortunately the MFEM API isn't quite as nice as to allow that.
`SetPreconditioner()` is not a member of `mfem::Solver`, but of `mfem::IterativeSolver` (base class of `mfem::CGSolver` and `mfem::GMRESSolver`) and `mfem::HyprePCG`, `mfem::HypreGMRES` and `mfem::HypreFGMRES`.
I've given it a try nonetheless, the best I could come up with is in the last commit, 1060460.
Feel free to amend and push if you come up with better ideas.

nmnobre committed:
Refactor preconditioner setting/checking logic (#30660)

nmnobre committed:
Add new option to control hypre AMG errors (#30660)

lindsayad committed:
Use unique_ptr for MFEMSolverBase::_solver (#30660)

And some other minor changes through the MFEM Solver classes

lindsayad commented:
@nmnobre let me know how you feel about the most recent commit

nmnobre commented on framework/src/mfem/solvers/MFEMSolverBase.C:
```diff
@@ -14,8 +14,44 @@ MFEMSolverBase::validParams()
+        solver.SetPreconditioner(*hypre_pre);
+      else
+        mooseError("hypre solver preconditioners must themselves be hypre solvers");
+    else if constexpr (std::is_base_of_v<mfem::IterativeSolver, T>)
+      solver.SetPreconditioner(mfem_pre);
+    else
+      mooseAssert(
+          true,
+          "We should never get here because we only instantiate this method for iterative solvers");
```
```suggestion
    else
      solver.SetPreconditioner(mfem_pre);
```

Since, as you note, we're explicitly instantiating anyway, maybe we could just get rid of both assertions in this method...? The compiler will complain anyway if `SetPreconditioner` isn't a member of `T`.

The suggestion above should actually cover all cases, but one... (sigh)
`mfem::HypreAME` which takes an `mfem::HypreSolver` for preconditioner, but is not itself an `mfem::HypreSolver`. But again, if we ever implement that solver and instantiate this function, the compiler will complain `mfem_pre` is of the wrong type, at which point we can amend line 37 with `|| T == mfem::HypreAME`.

lindsayad committed:
Remove assertions in setPreconditioner (#30660)

All our instantiations support setPreconditioner, which they must or else we would get a compile error anyway

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>

nmnobre committed:
Initialize hypre whenever we use hypre ¯\_(ツ)_/¯ (#30660)

We could initialize HYPRE every time we have an MFEMProblem,
as we assume MPI is present, and MFEM mandates HYPRE whenever
MPI is present. But, since it's safe to call mfem::Hypre::Init()
more than once, we do it lazily, only when any of our HYPRE
solvers is constructed.

nmnobre committed:
Fix memory leaks (#30660)

nmnobre commented:
@loganharbour, can we lift both the timeout and 5 maximum failures limits on the new valgrind test, just for the time being, 'cause it currently feels like a cat and mouse game xD

@lindsayad Feel free to add that segfault here if you'd like, we're going to beat the record for the most bugs fixed in a single PR haha

nmnobre committed:
Fix submesh BCs (#30660)

I'm surprised this made it through CI...
With gcc 15 the tests fail without this patch.

nmnobre committed:
Simplify ApplyBC to not take a mesh (#30660)

nmnobre commented:
I think this is done! Just 3 spurious errors: 2 timeouts and 1 due to too big a stack frame (~2MiB stack allocations in [framework/contrib/mfem/fem/qinterp/det.cpp:252](https://github.com/mfem/mfem/blob/9f945a5863407bdb5e1212465687ba09a49bbd7e/fem/qinterp/det.cpp#L252-L254) which confuse Valgrind). Can add some `valgrind = none`, or `valgrind = heavy`, or `max_time = ...`, but I'll let you decide.

![giphy](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExd25zbXJwc3YxMjRraWViZXFidHNrZWpsdHg5Nml1eTdneTdzc2RmMCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/bL0Taf2QKNJD2/giphy.gif)

lindsayad commented:
I think `valgrind = none` for the stack frame failure and `valgrind = heavy` for the timeouts. At some point we'll probably want to start splitting out tests from the `mfem/kernels `directory because it just keeps growing

nmnobre committed:
Appropriate mark time- or stack-heavy valgrind tests (#30660)
